jobs/gtfs-rt-parser-v2/gtfs_rt_parser.py

```diff
@@ -0,0 +1,740 @@
+
+    # https://github.com/MobilityData/gtfs-realtime-validator/issues/92
+    # try:
+    #     os.remove(os.path.join(dst_path, "areas.txt"))
+    # except FileNotFoundError:
+    #     pass
+
+    return actual_dst_path
+
+
+def execute_rt_validator(
+    gtfs_file: str, rt_path: str, jar_path: Path, verbose=False, pbar=None
+):
+    log(f"validating {rt_path} with {gtfs_file}", fg=typer.colors.MAGENTA, pbar=pbar)
+
+    args = [
+        "java",
+        "-jar",
+        str(jar_path),
+        "-gtfs",
+        gtfs_file,
+        "-gtfsRealtimePath",
+        rt_path,
+        "-sort",
+        "name",
+    ]
+
+    log(f"executing rt_validator: {' '.join(args)}", pbar=pbar)
+    subprocess.run(
+        args,
+        capture_output=True,
+    ).check_returncode()
+
+
+def validate_and_upload(
+    fs,
+    jar_path: Path,
+    dst_path_rt: str,
+    tmp_dir: str,
+    hour: RTHourlyAggregation,
+    verbose: bool = False,
+    pbar=None,
+) -> List[RTFileProcessingOutcome]:
+    first_extract = hour.extracts[0]
+    try:
+        gtfs_zip = download_gtfs_schedule_zip(
+            fs,
+            schedule_extract=get_schedule_extract_for_rt_validation(first_extract),
+            dst_path=tmp_dir,
+            pbar=pbar,
+        )
+    except FileNotFoundError:
+        raise ScheduleDataNotFound(f"no schedule data found for {first_extract}")
+
+    execute_rt_validator(
+        gtfs_zip,
+        dst_path_rt,
+        jar_path=jar_path,
+        verbose=verbose,
+        pbar=pbar,
+    )
+
+    records_to_upload = []
+    outcomes = []
+    for extract in hour.extracts:
+        results_path = os.path.join(
+            dst_path_rt, extract.timestamped_filename + ".results.json"
+        )
+        try:
+            with open(results_path) as f:
+                records = json.load(f)
+        except FileNotFoundError as e:
+            msg = f"WARNING: no validation output file found in {results_path}"
+            if verbose:
+                log(
+                    msg,
+                    fg=typer.colors.YELLOW,
+                    pbar=pbar,
+                )
+            outcomes.append(
+                RTFileProcessingOutcome(
+                    step=hour.step,
+                    success=False,
+                    exception=e,
+                    extract=extract,
+                )
+            )
+            continue
+
+        records_to_upload.extend(
+            [
+                {
+                    # back and forth so we can use pydantic serialization
+                    "metadata": serialize_pydantic_model_for_bigquery(extract),
+                    **record,
+                }
+                for record in records
+            ]
+        )
+
+        outcomes.append(
+            RTFileProcessingOutcome(
+                step=hour.step,
+                success=True,
+                extract=extract,
+            )
+        )
+
+    upload_if_records(
```

Comment:
if this throws an exception/fails, is it possible that we end up with no outcome for this input record?

Reply:
Not sure I love that TBH, I feel like I'd want the outcomes to be assigned only after we've actually done the whole action? So maybe at least change the order? I guess it shouldn't matter because it won't return the outcome if this fails but still, it feels wrong. I guess that wouldn't make a functional difference though so... 🤷

Reply:
Right, I think we don't have to worry about saving outcomes when we hit an unhandled exception i.e. failed to upload our final combined file. I've been trying to have outcomes only reflected handled exceptions on a per-input basis, rather than overall processing errors that will probably cause us to retry the job anyways.